main.nf

```diff
@@ -1,144 +1,189 @@
-        return 'unknown'
-    }
+bams = Channel.fromPath(params.ubam_dir + '/*.bam', checkIfExists: true)
+       .map{it -> tuple(it.baseName, it)}
+
+genome = params.genome
+params.reference_list = params.genomes[genome]
+genome_fa = Channel.value(params.reference_list.genome_fa)
+genome_fai = Channel.value("${params.reference_list.genome_fa}.fai")
+
+def checkFileSize (path) {
+    return path.toFile().length() >= 200   // Minimum size in bytes for a read file to be considered valid
 }
 
 
 workflow {
     main:
-        placeholder_r2 = touchFile( "placeholder.r2.fastq" )
 
-        // if reference is not indexed, index it.
-        if (!file(params.path_to_genome_fasta).exists()) {
-            println "Workflow failed: Genome file does not exist."
-            System.exit(1)  // Exit with a custom status code
+        passed_bams = bams.filter { library, bam -> checkFileSize(bam) }
+        failed_bams = bams.filter { library, bam -> !checkFileSize(bam) }
+        failed_library_names = failed_bams.map { library, bam -> library }
+
+        ////////// Filter libraries with insufficient reads //////////
+        failed_library_names.collect().subscribe { names ->
+            if (names.size() > 0) {
+                def joined_names = names.join('<br>')
+                sendMail {
+                    to params.email
+                    subject 'File Read Check'
+                    body """
+                    <html>
+                      <body>
+                        <p>The following libraries:<br> <strong>${joined_names}</strong> do not have enough reads. <br> Continuing with remaining libraries. </p>
+                      </body>
+                    </html>
+                    """
+                }
+            }
         }
 
-        genome_index_ch = bwa_index()
-
-        reads = Channel
-          .fromPath(params.input_glob)
-          .filter { !(it.name ==~ /^.*2\.fastq.*$/) }
-          .map { input_file ->
-            def fileType = detectFileType(input_file)
-            def read1File = input_file
-            def read2File = placeholder_r2.val  // val blocks until the value is available
-            if (fileType == 'fastq_paired_end') {
-                read2File = replaceReadNumber(input_file.toString())
-           }
-            if (read1File.toString() == read2File.toString()) {
-                log.error("Error: Detected paired-end file with read1: ${read1File} but no read2. What is different in the file name?")
-                throw new IllegalStateException("Invalid paired-end file configuration")
+        ///////// Trim, align and mark duplicates //////////
+        fastp( passed_bams )
+        if (params.single_end) {
+            fastq_chunks = fastp.out.trimmed_fastq
+            .flatMap { library, fq_files ->                 
+                def fq_list = fq_files instanceof List ? fq_files : [fq_files]
```

Comment:
i'd like to play around with this a bit to see if we can't find something more elegant, but in the meantime, if it works then it's fine.